hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/fsck/ReconSafeModeMgrTask.java

```diff
@@ -0,0 +1,121 @@
+import org.apache.hadoop.hdds.protocol.DatanodeDetails;
+import org.apache.hadoop.hdds.scm.container.ContainerID;
+import org.apache.hadoop.hdds.scm.container.ContainerInfo;
+import org.apache.hadoop.hdds.scm.container.ContainerManager;
+import org.apache.hadoop.hdds.scm.node.states.NodeNotFoundException;
+import org.apache.hadoop.ozone.recon.scm.ReconNodeManager;
+import org.apache.hadoop.ozone.recon.scm.ReconSafeModeManager;
+import org.apache.hadoop.ozone.recon.tasks.ReconTaskConfig;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.TimeUnit;
+
+import static org.apache.hadoop.hdds.HddsConfigKeys.HDDS_HEARTBEAT_INTERVAL;
+import static org.apache.hadoop.hdds.HddsConfigKeys.HDDS_HEARTBEAT_INTERVAL_DEFAULT;
+
+/**
+ * Class that scans the list of containers and keeps track if
+ * recon warm up completed and it emits safe node.
+ */
+public class ReconSafeModeMgrTask {
+
+  private static final Logger LOG =
+      LoggerFactory.getLogger(ReconSafeModeMgrTask.class);
+
+  private ContainerManager containerManager;
+  private ReconNodeManager nodeManager;
+  private ReconSafeModeManager safeModeManager;
+  private List<DatanodeDetails> allNodes;
+  private List<ContainerInfo> containers;
+  private OzoneConfiguration ozoneConfiguration;
+  private final long interval;
+  private final long dnHBInterval;
+
+  public ReconSafeModeMgrTask(
+      ContainerManager containerManager,
+      ReconNodeManager nodeManager,
+      ReconSafeModeManager safeModeManager,
+      ReconTaskConfig reconTaskConfig,
+      OzoneConfiguration ozoneConfiguration) {
+    this.safeModeManager = safeModeManager;
+    this.containerManager = containerManager;
+    this.nodeManager = nodeManager;
+    this.allNodes = nodeManager.getAllNodes();
+    this.containers = containerManager.getContainers();
+    this.ozoneConfiguration = ozoneConfiguration;
+    interval = reconTaskConfig.getSafeModeWaitThreshold().toMillis();
+    dnHBInterval = ozoneConfiguration.getTimeDuration(HDDS_HEARTBEAT_INTERVAL,
+        HDDS_HEARTBEAT_INTERVAL_DEFAULT, TimeUnit.MILLISECONDS);
+  }
+
+  public synchronized void start() {
+    long timeElapsed = 0L;
+    try {
+      checkForReconToEmitSafeNode();
+      while (safeModeManager.getInSafeMode() && timeElapsed <= interval) {
+        wait(dnHBInterval);
+        timeElapsed += dnHBInterval;
+        allNodes = nodeManager.getAllNodes();
+        containers = containerManager.getContainers();
+        checkForReconToEmitSafeNode();
+      }
+      if (safeModeManager.getInSafeMode()) {
+        safeModeManager.setInSafeMode(false);
+      }
+    } catch (Throwable t) {
+      LOG.error("Exception in Missing Container task Thread.", t);
+      if (t instanceof InterruptedException) {
+        Thread.currentThread().interrupt();
+      }
+    }
+  }
+
+  private void checkForReconToEmitSafeNode()
```

Comment:
Do you mean `checkForReconToExitSafeNode`?

Reply:
@smengcl - Ok, this has been renamed to "tryReconExitSafeMode". Pls re-review.